Sources/NIOExtras/QuiescingHelper.swift

```diff
@@ -51,30 +53,70 @@ private final class ChannelCollector {
     func channelAdded(_ channel: Channel) throws {
         self.eventLoop.assertInEventLoop()
 
-        guard self.lifecycleState != .shutdownCompleted else {
+        switch self.lifecycleState {
+        case .upAndRunning(var openChannels, let serverChannel):
+            openChannels[ObjectIdentifier(channel)] = channel
+            self.lifecycleState = .upAndRunning(openChannels: openChannels, serverChannel: serverChannel)
+
+        case .shuttingDown(var openChannels, let fullyShutdownPromise):
+            openChannels[ObjectIdentifier(channel)] = channel
+            channel.eventLoop.execute {
+                channel.pipeline.fireUserInboundEventTriggered(ChannelShouldQuiesceEvent())
+            }
+            self.lifecycleState = .shuttingDown(openChannels: openChannels, fullyShutdownPromise: fullyShutdownPromise)
+
+        case .shutdownCompleted:
             channel.close(promise: nil)
             throw ShutdownError.alreadyShutdown
         }
-
-        self.openChannels[ObjectIdentifier(channel)] = channel
     }
 
     private func shutdownCompleted() {
         self.eventLoop.assertInEventLoop()
-        assert(self.lifecycleState == .shuttingDown)
 
-        self.lifecycleState = .shutdownCompleted
-        self.fullyShutdownPromise?.succeed(())
+        switch self.lifecycleState {
+        case .upAndRunning:
+            preconditionFailure("This can never happen because we transition to shuttingDown first")
+
+        case .shuttingDown(_, let fullyShutdownPromise):
+            self.lifecycleState = .shutdownCompleted
+            fullyShutdownPromise.succeed(())
+
+        case .shutdownCompleted:
+            preconditionFailure("We should only complete the shutdown once")
+        }
     }
 
     private func channelRemoved0(_ channel: Channel) {
         self.eventLoop.assertInEventLoop()
-        precondition(self.openChannels.keys.contains(ObjectIdentifier(channel)),
-                     "channel \(channel) not in ChannelCollector \(self.openChannels)")
 
-        self.openChannels.removeValue(forKey: ObjectIdentifier(channel))
-        if self.lifecycleState != .upAndRunning && self.openChannels.isEmpty {
-            shutdownCompleted()
+        switch self.lifecycleState {
+        case .upAndRunning(var openChannels, let serverChannel):
+            precondition(
+                openChannels.keys.contains(ObjectIdentifier(channel)),
```

Comment:
Minor performance nit, but this precondition can be handled by the return value of `removeValue`, and avoid hashing the object identifier twice. Same note for the state below.